beanstalkd-put.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "BEANSTALKD\-PUT" "1" "March 2014" "beanstalkd-put 1.0" ""
.
.SH "NAME"
\fBbeanstalkd\-put\fR \- put a job into a beanstalkd tube
.
.SH "SYNOPSIS"
\fBbeanstalkd\-put\fR [\fIoptions\fR] \fI<job\-body\fR>
.
.SH "DESCRIPTION"
\fBbeanstalkd\-put\fR is a simple CLI utility to insert jobs into beanstalkd tubes\. The job body can either be specified on the command line or it can be read from \fBstdin\fR\.
.
.SH "OPTIONS"
.
.TP
\fB\-t\fR \fItube\fR, \fB\-\-tube\fR=\fItube\fR
The beanstalkd tube to put the job into\.
.
.TP
\fB\-s\fR \fIserver\fR, \fB\-\-server\fR=\fIserver\fR
The beanstalkd server hostname\.
.
.TP
\fB\-p\fR \fIport\fR, \fB\-\-port\fR=\fIport\fR
The beanstalkd server port number\.
.
.TP
\fB\-P\fR \fIpriority\fR, \fB\-\-priority\fR=\fIpriority\fR
The job priority\. Lower numbers are higher priority, thus maximum priority is 0 and minumum priority is 4294967295 (UINT32_MAX)\.
.
.TP
\fB\-D\fR \fIdelay\fR, \fB\-\-delay\fR=\fIdelay\fR
The number of seconds before the job becomes available to processors\.
.
.TP
\fB\-T\fR \fIttr\fR, \fB\-\-ttr\fR=\fIttr\fR
The number of seconds to allow a processor to complete the task without having to \fBtouch\fR the job to request more time\. A value of 0 gets silently converted into a 1 by the server, according to the beanstalkd 1\.3 protocol document \fIhttps://github\.com/kr/beanstalkd/tree/v1\.3/doc/protocol\.txt#L142\fR
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Usage help\.
.
.SH "JOB BODY"
If the "job\-body" is specified as "\-" then the body is read from stdin\.
.
.SH "OUTPUT"
After successfully putting the job into the tube, the job ID is printed to stdout\. The ID isn\'t surrounded by anything and is therefore easily captured for use in a script\.
.
.SH "EXAMPLES"
Put "Hello World" in the default tube:
.
.IP "" 4
.
.nf

$ beanstalkd\-put "Hello World"
.
.fi
.
.IP "" 0
.
.P
Pipe "Hello World" into the "howdy" tube:
.
.IP "" 4
.
.nf

$ echo "Hello World" | beanstalkd\-put \-\-tube=howdy \-
.
.fi
.
.IP "" 0
.
.P
Capture the job ID for display in a message:
.
.IP "" 4
.
.nf

$ JID=$(beanstalkd\-put "Hello World")
$ echo "The greeting was queued as job number $JID"
.
.fi
.
.IP "" 0
.
.SH "EXIT CODES"
There are three basic errors:
.
.P
1 \- Error while parsing CLI options\.
.
.P
2 \- Error communicating to beanstalkd\.
.
.P
3 \- Error submitting job body to beanstalkd\.
.
.SH "AUTHOR"
\fBJames Cook\fR \fIbonkabonka@gmail\.com\fR\.


debian/changelog
beanstalkd-put (1.0-1) unstable; urgency=low

  * Initial release.

 -- James Cook <bonkabonka@gmail.com>  Fri, 14 Mar 2014 15:23:31 -0400


beanstalkd-put.c
#include <getopt.h>
#include <libgen.h>
#include <limits.h>
#include <stdio.h>
#include <string.h>

#include "beanstalk.h"

#define MAX_JOB_SIZE_KEY "max-job-size: "

char *tube_name = "default";
char *server_host = "localhost";
int server_port = 11300;
// Priority 0 is max, UINT32_MAX is lowest
// See: https://github.com/kr/beanstalkd/blob/v1.3/doc/protocol.txt#l135
uint32_t job_priority = UINT32_MAX;
uint32_t job_delay = 0;
uint32_t job_ttr = 120;

void display_help(char *prog)
{
	fprintf(stderr, "Usage:\n");
	fprintf(stderr, "  %s [options] <job body>\n\n", basename(prog));
	fprintf(stderr, "Options:\n");
	fprintf(stderr, "  -t, --tube      <tube>      beantsalk tube to put the job into (%s)\n", tube_name);
	fprintf(stderr, "  -s, --server    <server>    beanstalk server hostname (%s)\n", server_host);
	fprintf(stderr, "  -p, --port      <port>      beanstalk server port number (%u)\n", server_port);
	fprintf(stderr, "  -P, --priority  <priority>  job priority, 0=max (%u)\n", job_priority);
	fprintf(stderr, "  -D, --delay     <delay>     seconds before the job becomes available (%u)\n", job_delay);
	fprintf(stderr, "  -T, --ttr       <ttr>       seconds to give the job processor (%u)\n", job_ttr);
	fprintf(stderr, "  -h, --help                  display this help\n\n");
	fprintf(stderr, "If \"job body\" is \"-\", then the body content will be read from stdin.\n\n");
}

int process_args(int argc, char **argv)
{
	static struct option long_options[] = {
		{ "tube",     required_argument, 0, 't' },
		{ "server",   required_argument, 0, 's' },
		{ "port",     required_argument, 0, 'p' },
		{ "priority", required_argument, 0, 'P' },
		{ "delay",    required_argument, 0, 'D' },
		{ "ttr",      required_argument, 0, 'T' },
		{ "help",     no_argument,       0, 'h' },
		{ NULL,       0,                 0, 0   }
	};

	int c;

	while ((c = getopt_long(argc, argv, "t:s:p:P:D:T:h", long_options, NULL)) != -1) {
		switch (c) {
			case 't':
				tube_name = optarg;
				break;
			case 's':
				server_host = optarg;
				break;
			case 'p':
				server_port = atoi(optarg);
				break;
			case 'P':
				job_priority = atoi(optarg);
				break;
			case 'D':
				job_delay = atoi(optarg);
				break;
			case 'T':
				job_ttr = atoi(optarg);
				break;
			default:
				display_help(argv[0]);
				return -1;
		}
	}

	if (optind >= argc || optind + 1 < argc) {
		display_help(argv[0]);
		return -1;
	}

	return optind;
}

int main(int argc, char **argv) {
	unsigned char *yaml;
	unsigned char *job;
	unsigned char *max_job_size_key;
	uint64_t max_job_size = 0;
	uint64_t job_size = 0;

	int c = process_args(argc, argv);
	if (c < 1) {
		return 1;
	}

	int socket = bs_connect(server_host, server_port);
	if (socket == BS_STATUS_FAIL) {
		fprintf(stderr, "Unable to connect to beanstalkd on %s:%d\n", server_host, server_port);
		return 2;
	}

	if (bs_use(socket, tube_name) == BS_STATUS_FAIL) {
		fprintf(stderr, "Unable to use beanstalkd tube %s\n", tube_name);
		bs_disconnect(socket);
		return 2;
	}

	if (bs_stats(socket, (char **)&yaml) == BS_STATUS_FAIL) {
		fprintf(stderr, "Unable to get beanstalkd stats\n");
		bs_disconnect(socket);
		return 2;
	}

	if ((max_job_size_key = (unsigned char *)strstr((char *)yaml, MAX_JOB_SIZE_KEY)) == NULL) {
		fprintf(stderr, "Unable to determine beanstalkd's max job size\n");
		free(yaml);
		bs_disconnect(socket);
		return 2;
	}

	max_job_size = atoi((char *)max_job_size_key + strlen(MAX_JOB_SIZE_KEY));

	free(yaml);

	job_size = strlen(argv[c]);
	if (job_size > max_job_size) {
		fprintf(stderr, "Job too large to send through beanstalkd %ld > %ld\n", job_size, max_job_size);
		bs_disconnect(socket);
		return 3;
	}

	// Add an extra byte to ensure null termination and to detect stdin overflow
	job = (unsigned char *)calloc(max_job_size + 1, 1);
	memcpy(job, argv[c], job_size);

	if (job_size == 1 && job[0] == '-') {
		job_size = 0;

		do {
			c = read(STDIN_FILENO, job + job_size, (1 + max_job_size) - job_size);
			if (c < 0) {
				perror("reading from stdin");
				free(job);
				bs_disconnect(socket);
				return 3;
			}
			job_size += c;
		} while (c > 0);

		if (job_size > max_job_size) {
			fprintf(stderr, "Job too large to send through beanstalkd %ld > %ld\n", job_size, max_job_size);
			free(job);
			bs_disconnect(socket);
			return 3;
		} else if (job_size == 0) {
			fprintf(stderr, "No job read from stdin\n");
			free(job);
			bs_disconnect(socket);
			return 3;
		}
	}

	int64_t id = bs_put(socket, job_priority, job_delay, job_ttr, (char *)job, job_size);
	if (id == 0) {
		fprintf(stderr, "Error putting job into tube\n");
		free(job);
		bs_disconnect(socket);
		return 3;
	}

	free(job);

	printf("%ld\n", id);

	bs_disconnect(socket);

	return 0;
}
